### PR TITLE
avoid splatting penalty in chkstride1

### LINKED ISSUE #16414: Splatting overhead in chkstride1 is significant for small matrices
This diff significantly reduces the time to solve a 10x10 LU factorized system:

```diff
diff --git a/base/linalg.jl b/base/linalg.jl
index 1feb700..58199f4 100644
--- a/base/linalg.jl
+++ b/base/linalg.jl
@@ -176,9 +176,11 @@ else
 end
 
 # Check that stride of matrix/vector is 1
+chkstride1(a) = stride(a,1) == 1 || error("matrix does not have contiguous columns")
 
diff --git a/base/linalg/lapack.jl b/base/linalg/lapack.jl
index e768619..6b32f27 100644
--- a/base/linalg/lapack.jl
+++ b/base/linalg/lapack.jl
@@ -918,7 +918,9 @@ for (gels, gesv, getrs, getri, elty) in
         #      DOUBLE PRECISION   A( LDA, * ), B( LDB, * )
         function getrs!(trans::Char, A::StridedMatrix{$elty}, ipiv::StridedVector{BlasInt}, B::StridedVecOrMat{$elty})
             chktrans(trans)
-            chkstride1(A, B, ipiv)
+            chkstride1(A)
+            chkstride1(B)
+            chkstride1(ipiv)
             n = checksquare(A)
             if n != size(B, 1)
                 throw(DimensionMismatch("B has leading dimension $(size(B,1)), but needs $n"))
```

This comes purely from avoiding the call to the [splatted `chkstride!` definition](https://github.com/JuliaLang/julia/blob/89a1fc23bb3269284a7d6c698bfa23f7e2ad1b9c/base/linalg.jl#L179).

With this benchmark

```jl
function f()
    b = lufact!(rand(10,10))
    c = rand(10)
    @time for i in 1:10^6
        A_ldiv_B!(b, c)
    end
end
```

with my diff:

```jl
julia> f()
  0.260720 seconds (1000.00 k allocations: 15.259 MB)
```

without
```jl
julia> f(b)
  0.365229 seconds (1000.00 k allocations: 15.259 MB
```

@andreasnoack Is it reasonable to just use the one arg version? There will perhaps be a bit of extra line clutter but 50% extra time to solve a 10x10 system is excessive... I guess there are many other functions where this also applies.

### PR DESCRIPTION
**Pull request by KristofferC**
~~Not sure this is kosher but it is late here so I blame that in case this is stupid.. now sleep.~~

Updated to use @timholy's suggestion in https://github.com/JuliaLang/julia/pull/16416#issuecomment-219899148

Fixes #16414

### CODE REVIEW

**Comment by yuyichao** [comment]
May as well change the code to call single argument version.

**Comment by andreasnoack** [comment]
A macro is fine with me here. Having six `chkstride1` calls in a row doesn't look good.

File: base/linalg.jl [hunk 1]
```diff
@@ -176,10 +176,11 @@ else
...
 # Check that stride of matrix/vector is 1
-function chkstride1(A...)
-    for a in A
-        stride(a,1)== 1 || error("matrix does not have contiguous columns")
-    end
+@inline _chkstride1(a) = stride(a,1) == 1 || error("matrix does not have contiguous columns")
+
+macro chkstride1(A...)
+    @assert isa(A, Tuple)
+    return Expr(:block, [:(_chkstride1($(esc(A[i])))) for i in 1:length(A)]...)
```
Comment (yuyichao) [inline base/linalg.jl]: I guess one advantage is that you can combine the check and have only one `error()`

File: base/linalg/lapack.jl [hunk 2]
```diff
@@ -8,7 +8,7 @@ const liblapack = Base.liblapack_name
 import Base.@blasfunc
 
 import ..LinAlg: BlasFloat, Char, BlasInt, LAPACKException,
-    DimensionMismatch, SingularException, PosDefException, chkstride1, checksquare
+    DimensionMismatch, SingularException, PosDefException, @chkstride1, _chkstride1, checksquare
```
Comment (yuyichao) [inline base/linalg/lapack.jl]: Don't import `_chkstride1`, it can hide hygiene bugs.

**Comment by timholy** [comment]
I would try writing it something like this:
```jl
julia> @inline chkstride(A...) = _chkstride(true, A...)
chkstride (generic function with 1 method)

julia> @noinline _chkstride(ok::Bool) = ok || error("matrix does not have contiguous columns")
_chkstride (generic function with 1 method)

julia> @inline _chkstride(ok::Bool, A, B...) = _chkstride(ok & (stride(A, 1) == 1), B...)
_chkstride (generic function with 2 methods)

julia> foo(A, B, C) = chkstride(A, B, C)   # this will let the inlining expand
foo (generic function with 1 method)

julia> foo(zeros(2,2), zeros(2,2), sub(zeros(4,2), 1:2:4, :))
ERROR: matrix does not have contiguous columns
 in _chkstride at none:1
 in foo at none:1

julia> @code_warntype foo(zeros(2,2), zeros(2,2), sub(zeros(4,2), 1:2:4, :))
Variables:
  A::Array{Float64,2}
  B::Array{Float64,2}
  C::SubArray{Float64,2,Array{Float64,2},Tuple{StepRange{Int64,Int64},Colon},1}
  ##A#7089::Tuple{Array{Float64,2},Array{Float64,2},SubArray{Float64,2,Array{Float64,2},Tuple{StepRange{Int64,Int64},Colon},1}}
  ####B#7086#7090::Tuple{Array{Float64,2},SubArray{Float64,2,Array{Float64,2},Tuple{StepRange{Int64,Int64},Colon},1}}
  ######B#7084#7087#7091::Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{StepRange{Int64,Int64},Colon},1}}
  ########B#7083#7085#7088#7092::Tuple{}

Body:
  begin  # none, line 1:
      GenSym(0) = (Main.stride)(A::Array{Float64,2},1)::Int64
      GenSym(1) = (Main.stride)(B::Array{Float64,2},1)::Int64
      GenSym(2) = (Main.stride)(C::SubArray{Float64,2,Array{Float64,2},Tuple{StepRange{Int64,Int64},Colon},1},1)::Int64
      return (Main._chkstride)((Base.box)(Base.Bool,(Base.and_int)((Base.box)(Base.Bool,(Base.and_int)((Base.box)(Base.Bool,(Base.and_int)(true,GenSym(0) === 1::Bool)),GenSym(1) === 1::Bool)),GenSym(2) === 1::Bool)))::Bool
  end::Bool
```
Then you probably don't need the macro.

Tuples rule.

**Comment by KristofferC** [comment]
OK, that works too. I'll be honest, I have no intuition about the whole "split function into sort of recursive parts and sprinkle `@inline` and `@noinline` and it avoids splatting" but it works here so.

**Comment by timholy** [comment]
I agree it's confusing. The rule that seems to work is to pay attention to every function in the call hierarchy that either (1) issues a splat-call or (2) is implemented as a slurp. Once all of those are inlined, you should be fine. In any other case, you're not. Since all callers of `chkstride` seem to know the precise arguments they want to pass, this seems like a safe case.

**Comment by timholy** [comment]
Nice & simple now, thanks!

**Pull request merged**